Support zephyr for SOF on i.MX8ULP

TangleZ opened the pull request:
Add support for i.MX8ULP to build Sound Open Firmware with Zephyr OS on DSP.
The build target is i.MX8ULP. It has one Xtensa HiFi4 core, with 64 KB ITCM, and 64KB DTCM.

Causing the toolchain not be ready, so the code not been tested in local so far.

iuliana-prodan commented on boards/xtensa/nxp_adsp_imx8ulp/nxp_adsp_imx8ulp_defconfig:
```diff
@@ -0,0 +1,22 @@
+
+CONFIG_SOC_SERIES_NXP_IMX8ULP=y
+CONFIG_BOARD_NXP_ADSP_IMX8ULP=y
+
```
some of these config are not needed anymore - see a patch that clears the imx8 and imx8m config - https://github.com/zephyrproject-rtos/zephyr/pull/62974/commits/cab0759282fc96515fa93776d20469552a82eccf
can you please do the same for 8ulp?

iuliana-prodan commented on soc/xtensa/nxp_adsp/imx8ulp/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,50 @@
+
+config SOC_SERIES
+	string
+	default "imx8ulp"
```
same as the above - needs some clean-ul, see https://github.com/zephyrproject-rtos/zephyr/pull/62974/commits/cab0759282fc96515fa93776d20469552a82eccf

TangleZ replied:
Compared to the imx8 and imx8m, this is aligned with them in the main branch.

iuliana-prodan replied:
@TangleZ I also agree with @LaurentiuM1234 we should clean up the config and `memory.h` file.
Please use these commits as examples:
- https://github.com/zephyrproject-rtos/zephyr/pull/62974/commits/9a293e50cd9ee6a3b8d776f197f5dc206bd61bfb
- https://github.com/zephyrproject-rtos/zephyr/pull/62974/commits/1c531389d109f69c2de4c43350f3293749c1314d

TangleZ replied:
Updated.

iuliana-prodan commented on soc/xtensa/nxp_adsp/imx8ulp/include/soc/memory.h:
```diff
@@ -0,0 +1,212 @@
+				+ SRAM_DEBUG_SIZE + SRAM_EXCEPT_SIZE \
+				+ SRAM_STREAM_SIZE + SRAM_TRACE_SIZE)
+
+/* Heap section sizes for module pool */
```
These are nowhere used - the HEAP_* and SOF_STACK*

LaurentiuM1234 replied:
Don't add all of this unnecessary information unless you know you're going to be using it in the future. Please remove it if not.

iuliana-prodan commented on soc/xtensa/nxp_adsp/imx8ulp/include/soc/memory.h:
```diff
@@ -0,0 +1,212 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __INC_MEMORY_H
```
Change the include guard to the standard ZEPHYR_

TangleZ replied:
cleaned up the file.

iuliana-prodan commented on soc/xtensa/nxp_adsp/imx8ulp/include/soc/platform.h:
```diff
@@ -0,0 +1,21 @@
+/*
```
This is not needed anymore - it was specific to SOF, but not used anymore - please remove it

TangleZ replied:
Removed the file

iuliana-prodan commented:
Please, solve compliance checks: https://github.com/zephyrproject-rtos/zephyr/actions/runs/6469020912/job/17563186840?pr=63751

dbaluta commented on boards/xtensa/nxp_adsp_imx8ulp/Kconfig.defconfig:
```diff
@@ -0,0 +1,21 @@
+	bool
+	default y
+#	depends on DMA
+
```
This shouldn't be commented out.

TangleZ replied:
Correct this.

dbaluta commented on boards/xtensa/nxp_adsp_imx8ulp/Kconfig.defconfig:
```diff
@@ -0,0 +1,21 @@
+	bool
+	default y
+#	depends on DMA
+
```
Same here. IMX_EDMA really depends on DMA.

dbaluta commented on soc/xtensa/nxp_adsp/imx8ulp/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,30 @@
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 666000000 if XTENSA_TIMER
+
```
@TangleZ what is the DSP frequency for imx8ulp? Looking at src/platform/imx8ulp/lib/clk.c it should be 528 MHz so please fix SYS_CLOCK_HW_CYCLES_PER_SEC

TangleZ replied:
Correct it to 528000000

Commit:
boards: xtensa: adsp: add support for imx8ulp board

Add support for i.MX8ULP board by defining defconfig,
board configuration and device tree.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
dts/xtensa/nxp: Add dtsi for imx8ulp

Add file nxp_imx8ulp.dtsi for imx8ulp

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
soc: xtensa: adsp: add support for NXP ADSP for i.MX8ULP

Add support for i.MX8ULP target.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
west: sign: add support for NXP i.MX8ULP board

Add support for signing i.MX8ULP SOF with Zephyr images with rimage.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

LaurentiuM1234 commented on boards/xtensa/nxp_adsp_imx8ulp/Kconfig.defconfig:
```diff
@@ -0,0 +1,21 @@
+config BOARD
+	default "nxp_adsp_imx8ulp"
+
+config DUMMY_DMA
```
Why's this needed? We've no native dummy DMA driver ATM. On SOF side the dummy dma is compiled via zephyr/CMakeLists.txt based on the SoC.

LaurentiuM1234 commented on boards/xtensa/nxp_adsp_imx8ulp/Kconfig.defconfig:
```diff
@@ -0,0 +1,21 @@
+	default y
+	depends on DMA
+
+config IMX_EDMA
```
Same question here.

TangleZ replied:
Removed the config.

LaurentiuM1234 commented on boards/xtensa/nxp_adsp_imx8ulp/nxp_adsp_imx8ulp_defconfig:
```diff
@@ -0,0 +1,21 @@
+
+CONFIG_XTENSA_SMALL_VECTOR_TABLE_ENTRY=y
+
+CONFIG_MULTI_LEVEL_INTERRUPTS=n
```
Why's this needed? AFAIK this is disabled by default.

LaurentiuM1234 commented on boards/xtensa/nxp_adsp_imx8ulp/nxp_adsp_imx8ulp_defconfig:
```diff
@@ -0,0 +1,21 @@
+CONFIG_XTENSA_SMALL_VECTOR_TABLE_ENTRY=y
+
+CONFIG_MULTI_LEVEL_INTERRUPTS=n
+CONFIG_2ND_LEVEL_INTERRUPTS=n
```
Same here.

LaurentiuM1234 commented on boards/xtensa/nxp_adsp_imx8ulp/nxp_adsp_imx8ulp_defconfig:
```diff
@@ -0,0 +1,21 @@
+CONFIG_GEN_ISR_TABLES=y
+CONFIG_GEN_IRQ_VECTOR_TABLE=n
+
+CONFIG_XTENSA_RESET_VECTOR=y
```
Why's this needed here? You're already enabling this via CONFG_SOC_SERIES_NXP_IMX8ULP

LaurentiuM1234 commented on boards/xtensa/nxp_adsp_imx8ulp/nxp_adsp_imx8ulp_defconfig:
```diff
@@ -0,0 +1,21 @@
+
+CONFIG_XTENSA_RESET_VECTOR=y
+
+CONFIG_XTENSA_USE_CORE_CRT1=y
```
Same here.

LaurentiuM1234 commented on boards/xtensa/nxp_adsp_imx8ulp/nxp_adsp_imx8ulp_defconfig:
```diff
@@ -0,0 +1,21 @@
+
+CONFIG_XTENSA_USE_CORE_CRT1=y
+
+CONFIG_XTENSA_SMALL_VECTOR_TABLE_ENTRY=y
```
This could be implicitly selected by `CONFIG_SOC_SERIEX_NXP_IMX8ULP` unless this is board specific.

LaurentiuM1234 commented on soc/xtensa/nxp_adsp/imx8ulp/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,30 @@
+	string
+	default "nxp_imx8ulp_adsp"
+
+config SOC
```
This seems to lack NXP HAL binding support. Please fix this as it's required for the switch to the native drivers. Ideally, support for serial interface should be also added as it's going to be required for debugging.

iuliana-prodan replied:
@LaurentiuM1234 I agree with you on configs clean up or other unused variables, but for the nxp-hal binding and uart support I believe we can add it later - we have to check what board we can use from hal, what files we need for the dsp, etc.. it might take a while.

LaurentiuM1234 replied:
I agree that this is **very** time consuming, but unless there's a time constraint I would still suggest getting it out of the way now. This is because later on we may have time constraints and this may become a problem. Either way's fine with me I guess

LaurentiuM1234 commented on soc/xtensa/nxp_adsp/imx8ulp/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,30 @@
+config SYS_CLOCK_TICKS_PER_SEC
+	default 50000
+
+config DYNAMIC_INTERRUPTS
```
Should probably enable this at board level. Any more opinions on this?

dbaluta replied:
We should stick with the approach we had for previous enabled boards. The we can fix it later (tm).

LaurentiuM1234 commented on soc/xtensa/nxp_adsp/imx8ulp/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,30 @@
+config DYNAMIC_INTERRUPTS
+	default y
+
+config LOG
```
Same here.

TangleZ commented:
@iuliana-prodan @LaurentiuM1234 @dbaluta  Thanks for your review, i updated the patch according to the comments, please have a look when you free.

iuliana-prodan commented:
> @iuliana-prodan @LaurentiuM1234 @dbaluta Thanks for your review, i updated the patch according to the comments, please have a look when you free.

@TangleZ this looks good to me, but we need to have the CI passing. So, we need the toolchain, for 8ULP, a in Zephyr SDK package. But we are kind of stuck on this: https://github.com/zephyrproject-rtos/sdk-ng/pull/712

LaurentiuM1234 reviewed (commented):
Not sure I understand why `CONFIG_IMX_EDMA` and `CONFIG_DUMMY_DMA` are still kept. Otherwise, LGTM.

TangleZ commented:
> Not sure I understand why `CONFIG_IMX_EDMA` and `CONFIG_DUMMY_DMA` are still kept. Otherwise, LGTM.

I referred the patch https://github.com/zephyrproject-rtos/zephyr/pull/62974/files#diff-43afd304a697d5b80f682e1e3e0a2a4e0950d5fb8b2fec9c275e24d080576ab9 from you, and there are still keep EDMA and DUMMY config. So I guess it's ok for ulp.

LaurentiuM1234 commented:
> > Not sure I understand why `CONFIG_IMX_EDMA` and `CONFIG_DUMMY_DMA` are still kept. Otherwise, LGTM.
> 
> I referred the patch https://github.com/zephyrproject-rtos/zephyr/pull/62974/files#diff-43afd304a697d5b80f682e1e3e0a2a4e0950d5fb8b2fec9c275e24d080576ab9 from you, and there are still keep EDMA and DUMMY config. So I guess it's ok for ulp.

My bad, missed them. I've updated my PR to remove them as well.

TangleZ commented:
@iuliana-prodan  I noticed that PR https://github.com/zephyrproject-rtos/sdk-ng/pull/712 is already merged, but the CI still fails, have we missed anything?

dbaluta commented:
@TangleZ  we still need Zephyr to use the new SDK. As per @stephanosio instructions - these will be around on Wednesday!

iuliana-prodan commented:
> @iuliana-prodan I noticed that PR [zephyrproject-rtos/sdk-ng#712](https://github.com/zephyrproject-rtos/sdk-ng/pull/712) is already merged, but the CI still fails, have we missed anything?

@TangleZ we need a zephyr sdk release that includes the toolchain for 8ulp.

@stephanosio will there be a sdk soon?
I believe that if we don't have the CI passing for this PR, with 8ulp support, it will not be merged.

stephanosio commented:
> @stephanosio will there be a sdk soon?
> I believe that if we don't have the CI passing for this PR, with 8ulp support, it will not be merged.

Yes, Zephyr SDK 0.16.4 should be released by Wednesday. Integrating it to the Zephyr main CI will take longer though.

TangleZ commented:
@stephanosio How can I know if the SDK merged to Zephyr main CI or not? And should I retrigger the PR to run the CI test?
